rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,533 @@
+written to the Write-Ahead Log (WAL), the data or transaction is considered
+durably written to storage. However, as discussed in the "Corruption Review"
+section of this document, this assumption is only theoretical. In practice,
+corruption may only be discovered when recovering from a system failure or
+crash, which is when WAL entries are most critical.
+
+It is important to note that during normal LSM operation, the movement of
+entries from the WAL to the SSTable occurs in memory. The WAL is only used as a
+recovery mechanism in case of a catastrophic failure of the LSM. Therefore, the
+loss of committed WAL data can only occur when both of the following conditions are
+true.
+
+1. The LSM application exits before compaction can complete.
+2. Corruption of the WAL occurs.
+
+Although the probability of both conditions being true is very low, especially
+in the context of durable object storage, it is not impossible.
+
+To address this, I propose that the `DB::open()` function perform WAL
+recovery synchronously and invoke `DBOptions.on_corruption` when WAL corruption
+is detected. If `DBOptions.on_corruption` returns an error, the database open
+operation should be aborted, and an error should be returned from
+`DB::open()`. If `DBOptions.on_corruption` does not return an error,
+corrupt entries in the WAL should be skipped until a valid entry is found.
+
+The corruption error returned should be of a distinct type that can easily be identified
+to determine whether the error is due to non-transient corruption or a
+transient issue (such as network connectivity problems).
+
+##### During MemTable Flush
+If corruption is detected during a MemTable flush and `DBOptions.on_corruption`
+returns an error, the compaction task will abort, and the database will be
+closed. Subsequent calls to `get` or `put` on a closed database will result in
+a closed database error.
+
+If `DBOptions.on_corruption` does not return an error, the current MemTable
+flush will abort, but the database will remain available. MemTable flush
+attempts will continue until the corruption is resolved. Calls to `put` 
+will succeed until the max number of un-flushed key/value pair bytes is reached.
+
+> NOTE: Corruption encountered during MemTable flush is likely to be a corrupted
+> manifest file. If manifest files are corrupted, there is no safe way to recover
+> and continue normal operation. In the case where `DBOptions.on_corruption` does not
+> return an error, we entertained the possibility of reverting to a previous manifest
+> and accepting data loss. However, there is no guarantee a previous manifest exists.
+
+#### Handling SST Corruption During Compaction
+Compaction is an asynchronous process that may encounter corruption in Manifest
+files, SST files, or missing SST files during operation.
+
+If corruption is detected during compaction `DBOptions.on_corruption` is invoked and 
+compaction will abort. Future compaction runs will continue to be attempted until
```

Comment:
If `on_corruption` doesn't return an error, couldn't we allow the compactor to skip the corrupt data and continue on (unless it's a corrupt manifest)?